dvc/remote/gdrive/__init__.py

```diff
@@ -0,0 +1,239 @@
+from __future__ import unicode_literals
+
+import os
+import posixpath
+
+from funcy import cached_property
+from ratelimit import limits, sleep_and_retry
+from backoff import on_exception, expo
+
+from dvc.scheme import Schemes
+from dvc.path_info import CloudURLInfo
+from dvc.remote.base import RemoteBASE
+from dvc.config import Config
+from dvc.exceptions import DvcException
+from dvc.remote.gdrive.pydrive import (
+    RequestListFile,
+    RequestListFilePaginated,
+    RequestUploadFile,
+    RequestDownloadFile,
+)
+from dvc.remote.gdrive.utils import FOLDER_MIME_TYPE
+
+
+class GDriveURLInfo(CloudURLInfo):
+    @property
+    def netloc(self):
+        return self.parsed.netloc
+
+
+class RemoteGDrive(RemoteBASE):
+    scheme = Schemes.GDRIVE
+    path_cls = GDriveURLInfo
+    REGEX = r"^gdrive://.*$"
+    REQUIRES = {"pydrive": "pydrive"}
+
+    def __init__(self, repo, config):
+        super(RemoteGDrive, self).__init__(repo, config)
+        self.no_traverse = False
+        self.cached_dirs = {}
+        self.cached_ids = {}
+        self.path_info = self.path_cls(config[Config.SECTION_REMOTE_URL])
+        self.config = config
+        self.init_drive()
+
+    def init_drive(self):
+        if Config.SECTION_REMOTE_KEY_FILE not in self.config:
+            raise DvcException(
+                "Google Drive settings file path is missed from config. "
+                "Learn more at https://dvc.org/doc."
+            )
+        self.gdrive_credentials_path = self.config.get(
+            Config.SECTION_REMOTE_KEY_FILE
+        )
+        self.root_id = self.get_path_id(self.path_info, create=True)
+        self.cache_root_dirs()
+
+    @on_exception(expo, DvcException, max_tries=8)
+    @sleep_and_retry
+    @limits(calls=10, period=1)
+    def execute_request(self, request):
+        try:
+            result = request.execute()
+        except Exception as exception:
+            if "Rate Limit Exceeded" in str(exception):
+                raise DvcException("API usage rate limit exceeded")
+            raise
+        return result
+
+    def list_drive_item(self, query):
+        list_request = RequestListFilePaginated(self.drive, query)
+        page_list = self.execute_request(list_request)
+        while page_list:
+            for item in page_list:
+                yield item
+            page_list = self.execute_request(list_request)
+
+    def cache_root_dirs(self):
+        for dir1 in self.list_drive_item(
+            "'{}' in parents and trashed=false".format(self.root_id)
+        ):
+            self.cached_dirs.setdefault(dir1["title"], []).append(dir1["id"])
+            self.cached_ids[dir1["id"]] = dir1["title"]
+
+    @cached_property
+    def drive(self):
+        from pydrive.auth import GoogleAuth
+        from pydrive.drive import GoogleDrive
+
+        if os.getenv("PYDRIVE_USER_CREDENTIALS_DATA"):
```

Comment:
could you please elaborate a little bit what is this logic for?

Reply:
`PYDRIVE_USER_CREDENTIALS_DATA` is supposed to be used by ci tools and be an encrypted user's login token to access user's GDrive disk space. Only some specially created test google account should be used on ci.

Reply:
is it documented somewhere? what specific problem does it solve? could you please elaborate a little bit more or share a link?

Reply:
so far, I've placed brief explanation of this in description of this PR:

> Environment variable PYDRIVE_USER_CREDENTIALS_DATA can be used to store user account's token data to skip manual login action. Should be useful for automated tested, but the actual value of PYDRIVE_USER_CREDENTIALS_DATA is supposed to be encrypted and secured from unauthorized usage.

to answer your questions:

> is it documented somewhere?
Not yet, but it should be.

> what specific problem does it solve?
test_data_cloud.py can run func tests on GDrive with automated login to some google user test account.

> could you please elaborate a little bit more or share a link?
sure, there are 2 types of Google's credentials DVC need to have to allow users to work with their GDrive:

1. "Google Project" should be created to get access to Google API. API usage rate limits are applied per "Google Project". We will ask each user to create their own "Google Project".
To access "Google Project" user should generate oauth2 credentials and save them in settings file formatted according to PyDrive's expectations ( tests/func/gdrive-settings.yaml  is an example of such settings file ).
Usually such credentials are shared by software provider publicly and there is no risk to leak some users private data.

2. Final DVC user on first action with GDrive remote will need to login into their Google Account, grant access rights to "Google Project" created on step 1, and copy/paste generated token into dvc's command line. After that PyDrive will create "./credentials.json" file with sensitive info to access user's GDrive.
`PYDRIVE_USER_CREDENTIALS_DATA` is workaround to avoid manual login into account. DVC reads value of `PYDRIVE_USER_CREDENTIALS_DATA` and write it to `.crendentials.json`. It's used later by PyDrive to access user's GDrive.

Reply:
Got it. Thanks, Max for sharing this info. Am I correct that this name `PYDRIVE_USER_CREDENTIALS_DATA` is our custom name? If it is the case, let's rename it into GDRIVE_USER_CREDENTIALS_DATA or something DVC-specific. Users don't care about `PYDRIVE` on the DVC interface level. Btw, is the format there PyDrive-specific?

Reply:
Renaming to `GDRIVE_USER_CREDENTIALS_DATA` sounds great!

> Btw, is the format there PyDrive-specific?

Yes, value of `GDRIVE_USER_CREDENTIALS_DATA` env var and content of `settings.yaml` are what is expected by PyDrive to communicate with google authorization library further and with official python google api client.

Reply:
👍 

one more question regarding the workflow:

> We will ask each user to create their own "Google Project".

It means going into Google Console (analog to AWS console), right? Is there any way to avoid this? Or any way to do this via command line?

Reply:
> It means going into Google Console (analog to AWS console), right? 
Yes

> Is there any way to avoid this? Or any way to do this via command line?
It seems there is no automated way to do all required things. In 2017 Google seems [released](https://cloud.google.com/blog/products/gcp/automating-project-creation-with-google-cloud-deployment-manager) `Google Cloud Deployment Manager` to automate creation of Google Projects, but usage samples say manual actions still are required as prerequisites to enable `Google Cloud Deployment Manager` API https://github.com/GoogleCloudPlatform/deploymentmanager-samples/tree/master/examples/v2/project_creation#prerequisites

Reply:
thanks! may be let's mention this link during when we start doing our docs.